### PR TITLE
[SPARK-46350][SS] Fix state removal for stream-stream join with one watermark and one time-interval condition

### PR DESCRIPTION
**Pull request by neilramaswamy**
### What changes were proposed in this pull request?

In these changes, we modify the stream-stream state removal logic to trigger and drop state for one side of a stream-stream join by looking for the existence of a watermark on the _other_ stream.

### Why are the changes needed?

Currently, we check whether to drop state for side `A` only  if there exists a watermark on side `A`. However, that is not a correct condition: state removal from one side of stream, in a stream-stream join, should happens when there exists a watermark on the _other_ side (and a time-interval condition).

If we have the time interval condition `L > R + 10`, the minimum value of `L` event-times that we need to hold onto is `min(R + 10)`, and we can drop all state from the left that is _less_ than that quantity. Distributing, we see that it is `min(R) + min(10)`, which is `min(R) + 10`. `min(R)` is the watermark of R. So, we should always check whether there exists a watermark on `R` to drop state from `L`.

This is not to say that our state watermark removal predicate is wrong. It's actually correct. The mistake in our implementation is that the predicate we run to decide whether to apply state removal is wrong. 

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Unit tests.

Without these changes (i.e. against current Spark master), all the state assertions about removal would fail, i.e. these lines: [here](https://github.com/apache/spark/pull/44323/files#diff-a271c249e6989ebaad63a491910597cd336049516e5b8e9a80f2bbd7368f458dR318), [here](https://github.com/apache/spark/pull/44323/files#diff-a271c249e6989ebaad63a491910597cd336049516e5b8e9a80f2bbd7368f458dR328), [here](https://github.com/apache/spark/pull/44323/files#diff-a271c249e6989ebaad63a491910597cd336049516e5b8e9a80f2bbd7368f458dR1085).

You might ask, why didn't we catch this before? Why didn't our current unit tests catch this? It's because all of our unit tests use two watermarks. We never before had a test with 2 streams that used just one watermark, even though in our documentation we say that watermarking both sides if "optional". 

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala [hunk 1]
```diff
@@ -218,11 +223,27 @@ object StreamingSymmetricHashJoinHelper extends Logging {
...
+          .toSeq
+
+        if (stateWatermarkAttributeSeq.size == 1) {
+          val expr = watermarkExpression(Some(stateWatermarkAttributeSeq.head), stateValueWatermark)
+          expr.map(JoinStateValueWatermarkPredicate.apply _)
+        } else {
+          // This should never happen, since if we have two attributes from the condition
+          // and one watermark (which we have, since isWatermarkDefinedOnInput is true),
+          // we should have one state watermark.
+          None
```
Comment (WweiL) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala:245]: if the schema is "v1, v2, ts" for both side of streams. And join condition is 
`l.v1 = r.v1 AND l.v2 = r.v2 AND r.ts > l.ts + INTERVAL 10 seconds" 
this would break right?

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingJoinSuite.scala [hunk 2]
```diff
@@ -257,6 +257,74 @@ class StreamingInnerJoinSuite extends StreamingJoinSuite {
...
+      .withColumn("right_value", $"_1")
+      .withColumn("right_time", timestamp_seconds($"_2"))
+
+    // Our condition is R > L + 10 and we have a watermark on the left, so state
+    // should be dropped on the right
+    val joined = leftStream
+      .join(
+        rightStream,
+        expr("left_value = right_value AND right_time > left_time + interval 10 seconds"),
+        "inner"
```
Comment (WweiL) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingJoinSuite.scala:283]: is the fix only valid with one time-interval condition? what if adding right_time < left_time - interval 10 seconds?

Comment (neilramaswamy) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingJoinSuite.scala:283]: Yes, it's also valid, but you would need two watermarks in that case:

https://github.com/apache/spark/pull/44323/files#diff-a271c249e6989ebaad63a491910597cd336049516e5b8e9a80f2bbd7368f458dR478

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala (hunk 1 above)
Comment (neilramaswamy) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala:245]: So, the `stateWatermarkAttributeSeq` should only contain two attributes total from `attributesInCondition`, since `attributesInCondition` is something like `L > R + 10`.

However, there could be more than 2 attributes if the user did `L > R1 + R2 + 10`. But I don't think we support that.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala [hunk 3]
```diff
@@ -198,31 +198,52 @@ object StreamingSymmetricHashJoinHelper extends Logging {
     val joinKeyOrdinalForWatermark: Option[Int] = findJoinKeyOrdinalForWatermark(
       leftKeys, rightKeys)
 
+    // Returns a predicate that drops data less than the state watermark.
+    // oneSideInputAttributes is the side from which state will be dropped, while
+    // otherSideInputAttributes is the side on which a watermark must be defined.
```
Comment (neilramaswamy) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala:203]:
```suggestion
```

**Review by rangadi — COMMENTED** [review]
Could you reduce changes by avoiding rename?
The tests look good.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala [hunk 4]
```diff
@@ -198,31 +198,50 @@ object StreamingSymmetricHashJoinHelper extends Logging {
     val joinKeyOrdinalForWatermark: Option[Int] = findJoinKeyOrdinalForWatermark(
       leftKeys, rightKeys)
 
+    // Returns a predicate that drops data less than the state watermark.
     def getOneSideStateWatermarkPredicate(
-        oneSideInputAttributes: Seq[Attribute],
-        oneSideJoinKeys: Seq[Expression],
-        otherSideInputAttributes: Seq[Attribute]): Option[JoinStateWatermarkPredicate] = {
-      val isWatermarkDefinedOnInput = oneSideInputAttributes.exists(_.metadata.contains(delayKey))
+        stateRemovalSideAttributes: Seq[Attribute],
```
[the remark this answers is not in the document]
Comment (neilramaswamy) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala:203]: They're not strictly needed, but a possible reason why this mistake was made in the first place was that you're working with "one side" and "other side", which means nothing if you don't have code context.

If anything, this should make this PR easier to read. But if you feel strongly, I can change it back.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingJoinSuite.scala [hunk 5]
```diff
@@ -257,6 +257,78 @@ class StreamingInnerJoinSuite extends StreamingJoinSuite {
     )
   }
 
+  test("wei comment") {
+
+  }
```
Comment (neilramaswamy) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingJoinSuite.scala:262]:
```suggestion
```

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingJoinSuite.scala [hunk 6]
```diff
@@ -257,6 +257,78 @@ class StreamingInnerJoinSuite extends StreamingJoinSuite {
...
+        expr("left_value = right_value AND right_time > left_time + interval 10 seconds"),
+        "inner"
+      )
+      .select($"left_value", $"left_time".cast("int"), $"right_time".cast("int"))
+
+    testStream(joined)(
+      StartStream(),
+      AddData(leftSource, (4, 100)),
+      AddData(rightSource, (3, 105)),
+      CheckNewAnswer(),
```
[the remark this answers is not in the document]
Comment (neilramaswamy) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingJoinSuite.scala:292]: It asserts that the data passed in (nothing) is equal to the Spark answer:

https://github.com/apache/spark/blob/c417c757907e262f78af5a84b8358e253d0d0954/sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamTest.scala#L774

**Comment by neilramaswamy** [comment]
As I was looking through this code some more, I realized that `getStateWatermark`, which is used to produce the output watermark, also [has this same incorrect logic](https://github.com/apache/spark/blob/c417c757907e262f78af5a84b8358e253d0d0954/sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala#L157) for generating the watermark. This makes me think that one time interval condition with multiple stateful operators is also broken, since we won't return _any_ watermark, preventing propagation.

I'll have to confirm if this is true, and then open another ticket if so. But just wanted to make reviewers aware that I am aware of this.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala [hunk 7]
```diff
@@ -198,18 +198,23 @@ object StreamingSymmetricHashJoinHelper extends Logging {
...
+      val isWatermarkDefinedOnInput = watermarkAttribute.isDefined
       val isWatermarkDefinedOnJoinKey = joinKeyOrdinalForWatermark.isDefined
 
       if (isWatermarkDefinedOnJoinKey) { // case 1 and 3 in the StreamingSymmetricHashJoinExec docs
         val keyExprWithWatermark = BoundReference(
           joinKeyOrdinalForWatermark.get,
           oneSideJoinKeys(joinKeyOrdinalForWatermark.get).dataType,
-          oneSideJoinKeys(joinKeyOrdinalForWatermark.get).nullable)
+          oneSideJoinKeys(joinKeyOrdinalForWatermark.get).nullable
+        )
```
Comment (neilramaswamy) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala:217]:
```suggestion
          oneSideJoinKeys(joinKeyOrdinalForWatermark.get).nullable)
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala [hunk 8]
```diff
@@ -218,11 +222,27 @@ object StreamingSymmetricHashJoinHelper extends Logging {
           attributesToFindStateWatermarkFor = AttributeSet(oneSideInputAttributes),
           attributesWithEventWatermark = AttributeSet(otherSideInputAttributes),
           condition,
-          eventTimeWatermarkForEviction)
-        val inputAttributeWithWatermark = oneSideInputAttributes.find(_.metadata.contains(delayKey))
-        val expr = watermarkExpression(inputAttributeWithWatermark, stateValueWatermark)
-        expr.map(JoinStateValueWatermarkPredicate.apply _)
+          eventTimeWatermarkForEviction
+        )
```
Comment (rangadi) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala:226]: spurious change?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala [hunk 9]
```diff
@@ -218,11 +222,27 @@ object StreamingSymmetricHashJoinHelper extends Logging {
...
+
+        val attributesInCondition = AttributeSet(
+          condition.get.collect { case a: AttributeReference => a }
```
Comment (rangadi) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala:229]: add a comment. What are we doing here? I don't know why it is the right thing to do.

Comment (rangadi) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala:229]: Is every `Attribute` an `AttributeReference`? (I don't know much about this class hierarchy)

**Review by rangadi — COMMENTED** [review]
I need to understand better what we were doing before and what we are doing now.
May be we could have VC. I don't have much familiarity with this part of the code.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala (hunk 9 above)
Comment (neilramaswamy) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala:232]: Comment added.

No, the hierarchy is:

```
Attribute --> AttributeReference --> ColumnarAttributeReference
|
| --> PrettyAttribute (used only for debugging)
```

So it's safe to just use an `AttributeReference` in the collect. We also do this when we generate the state value watermark predicate: https://github.com/apache/spark/blob/f3a6ca9e2c476fe418679f82b0c5148d69d0fbda/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/StreamingJoinHelper.scala#L139

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala [hunk 10]
```diff
@@ -219,10 +222,35 @@ object StreamingSymmetricHashJoinHelper extends Logging {
...
+        )
+
+        // Construct an AttributeSet so that we can perform equality between attributes,
+        // which we do in the filter below.
+        val oneSideInputAttributeSet = AttributeSet(oneSideInputAttributes)
+
+        // oneSideInputAttributes could be [left_value, left_time], and we just
+        // want the attribute _in_ the time-interval condition.
+        val oneSideStateWatermarkAttributes = attributesInCondition.filter { a =>
+            oneSideInputAttributeSet.contains(a)
```
Comment (rangadi) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala:240]: What ensures this is the event-time attribute? Is this assured to be `left_time` mentioned in the comment?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala [hunk 11]
```diff
@@ -219,10 +222,35 @@ object StreamingSymmetricHashJoinHelper extends Logging {
...
+        }
+
+        // There should be a single attribute per side in the time-interval condition, so,
+        // filtering for oneSideInputAttributes as done above should lead us with 1 attribute.
+        if (oneSideStateWatermarkAttributes.size == 1) {
+          val expr = watermarkExpression(
+            Some(oneSideStateWatermarkAttributes.head), stateValueWatermark)
+          expr.map(JoinStateValueWatermarkPredicate.apply _)
+        } else {
+          // But, if there are more state watermark attributes, we can't solve the contraints,
```
Comment (rangadi) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala:250]: An example would be useful.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala [hunk 12]
```diff
@@ -219,10 +222,35 @@ object StreamingSymmetricHashJoinHelper extends Logging {
           attributesWithEventWatermark = AttributeSet(otherSideInputAttributes),
           condition,
           eventTimeWatermarkForEviction)
-        val inputAttributeWithWatermark = oneSideInputAttributes.find(_.metadata.contains(delayKey))
```
Comment (rangadi) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala:222]: Where is the equivalent of this statement? This looks like the event-time attribute from one side (left side in examples).

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala [hunk 13]
```diff
@@ -219,10 +222,35 @@ object StreamingSymmetricHashJoinHelper extends Logging {
...
+        // which we do in the filter below.
+        val oneSideInputAttributeSet = AttributeSet(oneSideInputAttributes)
+
+        // oneSideInputAttributes could be [left_value, left_time], and we just
+        // want the attribute _in_ the time-interval condition.
+        val oneSideStateWatermarkAttributes = attributesInCondition.filter { a =>
+            oneSideInputAttributeSet.contains(a)
+        }
+
+        // There should be a single attribute per side in the time-interval condition, so,
```
Comment (rangadi) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala:243]: Where is this part ensured? ('single attribute per side').

**Review by rangadi — COMMENTED** [review]
I understand this a bit more, but not fully. Could you use a concrete example? may be like the one @WweiL used in his comment.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala (hunk 10 above)
Comment (neilramaswamy) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala:240]: The interesting thing about this PR is that this _won't_ be the event-time attribute. If we only have one known watermarked column, we have to find the other attribute, and that other attribute is the one for which we do state removal. It does not necessarily need to be an event-time attribute.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala (hunk 12 above)
Comment (neilramaswamy) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala:222]: The code previously made the over-simplified assumption that there would be a watermark on one side, the state removal side. It would then use this assumption to find the attribute for state removal.

This assumption is not right. If you have the expression `L > R + 10` and `R` has a watermark, how do you find `L`? The approach I have chosen to take is taking the attributes (`attributesInCondition`), filtering for the attributes on the removal side (`oneSideInputAttributes`), and then taking that result. That would leave us with `R`.

Effectively, this line is equivalent to `oneSideStateWatermarkAttributes.head`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala (hunk 13 above)
Comment (neilramaswamy) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala:243]: I'm having a hard time finding exactly where we specify the grammar for the join `Expression`, but empirically, the grammar is something like: `ATTRIBUTE`, `COMPARATOR`, `ATTRIBUTE`, and then an `INTERVAL`. So, you can't specify more than 2 attributes without getting an exception.

We also make this assumption in higher up in `StreamingJoinHelper.getStateValueWatermark` as well: https://github.com/apache/spark/blob/f3a6ca9e2c476fe418679f82b0c5148d69d0fbda/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/StreamingJoinHelper.scala#L142

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala [hunk 14]
```diff
@@ -219,10 +222,41 @@ object StreamingSymmetricHashJoinHelper extends Logging {
...
-        expr.map(JoinStateValueWatermarkPredicate.apply _)
 
+        // If the condition itself is empty (for example, left_time < left_time + INTERVAL ...),
+        // then we will not have generated a stateValueWatermark.
+        if (stateValueWatermark.isEmpty) {
+          None
+        } else {
+          // For example, if the condition is of the form:
+          //    left_time > right_time + INTERVAL 30 MINUTES
+          // Then this extracts left_time and right_time.
```
Comment (rangadi) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala:233]: Is `condition` here only the time-interval part of the join condition? (e.g. consider 'A.id = B.id AND A.ts > B.ts + INTERNAL 30 MINUTES).

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala (hunk 12 above)
Comment (rangadi) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala:222]: > filtering for the 

Did you mean filtering out? 

> Effectively, this line is equivalent to oneSideStateWatermarkAttributes.head.

`This line`: Is that the line you removed?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala (hunk 14 above)
Comment (rangadi) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala:233]: I.e. can there be an assert here that `attributesInCondition` is exactly two timestamps, one on each side?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala (hunk 10 above)
Comment (rangadi) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala:245]: Can you give an example? 

> Is this assured to be left_time mentioned in the comment?

What about this part?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala [hunk 15]
```diff
@@ -219,10 +222,41 @@ object StreamingSymmetricHashJoinHelper extends Logging {
...
+          }
+
+          // There should be a single attribute per side in the time-interval condition, so,
+          // filtering for oneSideInputAttributes as done above should lead us with 1 attribute.
+          if (oneSideStateWatermarkAttributes.size == 1) {
+            val expr =
+              watermarkExpression(Some(oneSideStateWatermarkAttributes.head), stateValueWatermark)
+            expr.map(JoinStateValueWatermarkPredicate.apply _)
+          } else {
+            // This should never happen, since the grammar will ensure that we have one attribute
```
Comment (rangadi) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala:255]: > // This should never happen,

Why not assert here?

**Review by rangadi — APPROVED** [review]
Overall LGTM.
I don't have full visibility into exact context here. Some of the assertions and assumptions are not clear to me as a result. 
I do understand the core bug and the fix proposed.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala (hunk 12 above)
Comment (neilramaswamy) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala:222]: > Did you mean filtering out?

Well, we _want_ the attributes on `oneSideInputAttributes`, so it's not filtering out, it's filtering for. If you look at the new `.filter` condition, we are doing `oneSideInputAttributeSet.contains(a)`, so we only get back the attributes that are in `oneSideInputAttributes`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala (hunk 15 above)
Comment (neilramaswamy) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala:255]: Good question. I thought more about this, and I actually think that I might be wrong in the case of an edge-case we don't have in any of our tests: if the user does:

`left_time > right_time + m AND other_left_time > right_time + n`, there will be _three_ attributes in the condition. Then, `oneSideStateWatermarkAttributes.size` will be 2 (it will be `left_time` and `other_left_time`, neither of which are watermark attributes), and the condition that we need to return would be a conjunctive watermark predicate: `left_time <= watermark(right) + m AND other_left_time <= watermark(right) + n`.

We can remove state, but I'm pretty sure the current implementation in Spark master would fail. I need to check this. It might be out-of-scope for this PR.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala [hunk 16]
```diff
@@ -219,10 +222,41 @@ object StreamingSymmetricHashJoinHelper extends Logging {
...
+          // oneSideInputAttributes could be [left_value, left_time], and we just
+          // want the attribute _in_ the time-interval condition.
+          val oneSideStateWatermarkAttributes = attributesInCondition.filter { a =>
+            oneSideInputAttributeSet.contains(a)
+          }
+
+          // There should be a single attribute per side in the time-interval condition, so,
+          // filtering for oneSideInputAttributes as done above should lead us with 1 attribute.
+          if (oneSideStateWatermarkAttributes.size == 1) {
+            val expr =
```
Comment (sahnib) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinHelper.scala:251]: Discussed offline as well. This assumption does not seem to be correct. We actually need to find the partial join condition where the otherSide has eventTime attribute, and use that attribute to calculate watermark predicate. 

As an aside, it might be beneficial to combine this function with `getStateWatermark` as both of these have similar logic.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingJoinSuite.scala [hunk 17]
```diff
@@ -257,6 +257,75 @@ class StreamingInnerJoinSuite extends StreamingJoinSuite {
     )
   }
 
+
```
Comment (sahnib) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingJoinSuite.scala:260]: Lets also add a testcase for join condition where we compare eventTime and some other attribute (example - id)

See https://github.com/apache/spark/pull/44323/files#r1582300122 for context.

**Review by sahnib — COMMENTED** [review]
Thanks for making these changes. Left some comments/questions.
